File: smartmoneyserver/src/main/java/fpt/aptech/server/dto/planned/PlannedTransactionResponse.java
package fpt.aptech.server.dto.planned;

import lombok.Builder;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Builder
public record PlannedTransactionResponse(
    Integer id,

    // Ví
    Integer walletId,
    String  walletName,
    String  walletIcon,      // ✅ THÊM: Icon URL ví (Flutter hiển thị icon ví)

    // Danh mục
    Integer categoryId,
    String  categoryName,
    String  categoryIcon,
    Boolean categoryType, // false=Chi, true=Thu

    // Nợ (nullable)
    Integer debtId,
    String  debtPersonName,

    String     note,
    BigDecimal amount,

    // 1: Bill | 2: Recurring
    Integer planType,

    // Lịch lặp
    Integer repeatType,      // 0-4
    Integer repeatInterval,
    Integer repeatOnDayVal,  // bitmask tuần

    LocalDate beginDate,
    LocalDate nextDueDate,   // Flutter dùng để hiển thị "Lần tiếp theo: Thứ Hai 16/03/2026"
    LocalDate lastExecutedAt,
    LocalDate endDate,

    Boolean  active,
    LocalDateTime createdAt,

    // Mô tả lịch lặp cho Flutter hiển thị (Backend tính sẵn)
    // VD: "Lặp vào ngày 14, mỗi 2 tháng" | "Lặp mỗi T2, T4, T6 hàng tuần"
    String repeatDescription,

    // ══════════════════════════════════════════════════════════════════
    // ✅ BACKEND TÍNH SẴN — Cung cấp trạng thái & label rõ ràng cho UI
    // ══════════════════════════════════════════════════════════════════

    // Trạng thái hiển thị chính, quyết định màu sắc & hành động trên UI.
    // VD: "ACTIVE", "OVERDUE", "INACTIVE", "EXPIRED", "PAID"
    String displayStatus,

    // Label phụ, cung cấp thông tin chi tiết hơn cho người dùng.
    // VD: "Còn 3 lần", "Hết hạn 31/03/2026", "Đến hạn sau 5 ngày", "Quá hạn 2 ngày"
    String statusLabel,

    // Số lần lặp lại còn lại (chỉ có giá trị khi lặp theo số lần).
    // VD: 5, 4, 3, 2, 1
    Integer remainingCount,

    // Label ngày đến hạn đã format sẵn tiếng Việt.
    // VD: "Thứ Ba, 14 tháng 4 2026"
    String nextDueDateLabel
) {}

File: APP MOBILE/smart_money/lib/modules/planned/models/planned_transaction_response.dart
/// Response hiển thị thông tin giao dịch lặp lại.
/// Tương ứng: PlannedTransactionResponse.java (server)
class PlannedTransactionResponse {
  final int id;

  final int? walletId;
  final String? walletName;

  final int? categoryId;
  final String? categoryName;
  final String? categoryIcon;
  final bool? categoryType;

  final int? debtId;
  final String? debtPersonName;

  final String? note;
  final double amount;

  final int planType;

  final int? repeatType;
  final int? repeatInterval;
  final int? repeatOnDayVal;

  final DateTime? beginDate;
  final DateTime? nextDueDate;
  final DateTime? lastExecutedAt;
  final DateTime? endDate;

  final bool? active;
  final DateTime? createdAt;

  final String? repeatDescription;

  const PlannedTransactionResponse({
    required this.id,
    this.walletId,
    this.walletName,
    this.categoryId,
    this.categoryName,
    this.categoryIcon,
    this.categoryType,
    this.debtId,
    this.debtPersonName,
    this.note,
    required this.amount,
    required this.planType,
    this.repeatType,
    this.repeatInterval,
    this.repeatOnDayVal,
    this.beginDate,
    this.nextDueDate,
    this.lastExecutedAt,
    this.endDate,
    this.active,
    this.createdAt,
    this.repeatDescription,
  });

  factory PlannedTransactionResponse.fromJson(Map<String, dynamic> json) {
    return PlannedTransactionResponse(
      id: json['id'] as int,
      walletId: json['walletId'] as int?,
      walletName: json['walletName'] as String?,
      categoryId: json['categoryId'] as int?,
      categoryName: json['categoryName'] as String?,
      categoryIcon: json['categoryIcon'] as String?,
      categoryType: json['categoryType'] as bool?,
      debtId: json['debtId'] as int?,
      debtPersonName: json['debtPersonName'] as String?,
      note: json['note'] as String?,
      amount: (json['amount'] as num).toDouble(),
      planType: json['planType'] as int,
      repeatType: json['repeatType'] as int?,
      repeatInterval: json['repeatInterval'] as int?,
      repeatOnDayVal: json['repeatOnDayVal'] as int?,
      beginDate: json['beginDate'] != null
          ? DateTime.parse(json['beginDate'] as String)
          : null,
      nextDueDate: json['nextDueDate'] != null
          ? DateTime.parse(json['nextDueDate'] as String)
          : null,
      lastExecutedAt: json['lastExecutedAt'] != null
          ? DateTime.parse(json['lastExecutedAt'] as String)
          : null,
      endDate: json['endDate'] != null
          ? DateTime.parse(json['endDate'] as String)
          : null,
      active: json['active'] as bool?,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : null,
      repeatDescription: json['repeatDescription'] as String?,
    );
  }
}

